grader_systems/src/stage/valgrind.rs

```diff
@@ -0,0 +1,1123 @@
+        executable: String,
+        args: Vec<String>,
+        stdin: Option<String>,
+        timeout: Option<Duration>,
+    ) -> Self {
+        Self {
+            executor,
+            config,
+            executable,
+            args,
+            stdin,
+            timeout,
+        }
+    }
+
+    fn gen_cmd(&self, ws: &Path) -> Command {
+        let mut cmd = Command::new("valgrind");
+
+        cmd.add_arg(format!("--log-file={}", &self.config.log_file));
+
+        if let Some(v) = self.config.leak_check {
+            cmd.add_arg(format!("--leak-check={}", if v { "yes" } else { "no" }));
+        }
+
+        cmd.add_arg(format!("--error-exitcode={}", ERROR_EXITCODE));
+
+        if let Some(v) = self.config.malloc_fill {
+            cmd.add_arg(format!("--malloc-fill=0x{:02X}", v));
+        }
+
+        if let Some(v) = self.config.free_fill {
+            cmd.add_arg(format!("--free-fill=0x{:02X}", v));
+        }
+
+        if let Some(v) = &self.config.suppressions {
+            for supp in v {
+                cmd.add_arg(format!("--suppressions={}", supp));
+            }
+        }
+
+        cmd.add_arg("--");
+        cmd.add_arg(&self.executable);
+        cmd.add_args(&self.args);
+        if let Some(v) = &self.stdin {
+            cmd.set_stdin(StdinPipe::Path(v.into()))
+        }
+
+        cmd.set_timeout(self.timeout.unwrap_or(DEFAULT_TIMEOUT));
+        cmd.set_cwd(ws);
+
+        cmd
+    }
+
+    fn read_logfile(&self, ws: &Path) -> Result<String> {
+        let path = ws.join(&self.config.log_file);
+        if !path.exists() {
+            return Err(anyhow!(
+                "Could not find logfile at {:?}. Did Valgrind actually execute?",
+                path.to_str()
+            ));
+        }
+        let contents = read_file(&path)?;
+        if contents.trim().is_empty() {
+            return Err(anyhow!(
+                "Found empty valgrind log at {:?}. Something went wrong.",
+                path.to_str()
+            ));
+        }
+
+        // replace all absolute paths with basename
+        let re = Regex::new(r"(\W|^)(?:\/[^\/\s]+)+\/([^\/\s]+)\b")?;
+        let repl = re.replace_all(&contents, "$1$2");
+
+        Ok(repl.to_string())
+    }
+}
+
+#[async_trait]
+impl<E: ProcessExecutor> Executor for Valgrind<E> {
+    type Output = StageResult;
+
+    async fn run(&self, ws: &Path) -> Result<Self::Output> {
+        let mut sect = Section::new("Valgrind");
+        let mut results_sect = Section::new("Output:");
+        let mut run_updates = StatusUpdates::default();
+        debug!("running program");
+
+        let executable = ws.join(&self.executable);
+        if !executable.exists() {
+            return Err(anyhow!(
+                "Could not find student executable at {:?}",
+                executable.display()
+            ));
+        }
+
+        if let Some(v) = &self.config.suppressions {
+            for supp in v {
+                if !ws.join(&supp).exists() {
```

Comment:
Are the suppressions expected to be in the ws always? Are the graders expected to use the import option to import them? Or are they part of the basefiles? How do the suppressions get to the ws?

If they are in the test resource directory, then you will need to use a TestResourceLocator to "find" the actual path to the files.

Reply:
I think they should probably be basefiles. How would I access them in that case?

Reply:
Base files are something that are copied over to every assignment. Don't think that's the right choice. The suppressions can just be a normal test resource and as such can either be in the test_X directory or in static if used by multiple test cases. 

You can use a file finder to get those files. Check out the compare stage to see how those are used.

Reply:
Ah of course, since it's configurable. Will update that